src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.index.codec;
+
+import org.opensearch.index.codec.CodecServiceConfig;
+import org.opensearch.knn.index.codec.KNN87Codec.KNN87Codec;
+import org.apache.lucene.codecs.Codec;
+import org.opensearch.index.codec.CodecService;
+
+/**
+ * KNNCodecService to inject the right KNNCodec version
+ */
+public class KNNCodecService extends CodecService {
+
+    public KNNCodecService(CodecServiceConfig codecServiceConfig) {
+        super(codecServiceConfig.getMapperService(), codecServiceConfig.getLogger());
+    }
+
+    /**
+     * Return the custom k-NN codec that wraps another codec that a user wants to use for non k-NN related operations
+     *
+     * @param name of delegate codec.
+     * @return Latest KNN Codec built with delegate codec.
+     */
+    @Override
+    public Codec codec(String name) {
+        return new KNN87Codec(super.codec(name));
```

Comment:
I think we need some sort of factory here that will allow to select exact k-NN Codec implementation. I'm thinking of a use case for 2.0/lucene 9, we'll need to use Lucene90 codec for it, and probably will create new kNN wrapper class, something like KNN90Codec. Although it might be too much for this PR, we can do it in next one.

Reply:
Ok, I see, so essentially we change it with every next Lucene version but keep older codecs for bwc. Seems that this method (codec(String name)) is the factory method itself. Do we have our abstraction for knn codec apart from lucene abstract Codec class? It may give us more flexibility with testing, although I'm not sure it's that easy to pass our interface to the OpenSearch through plugin related service classes.

Reply:
Right, so we keep older Codecs so that older indices can be read by old Codecs via SPI. 

BWC is also why we need to version our Codecs (KNN87Codec). This is so that older indices can be read with old codecs.

I think this is something that could be investigated more in a future review.